TEE supplicant support

Deedone opened the pull request.

Deedone committed:
tests: drivers: optee: Unit tests for supplicant comms

Tests simple supplicant communication success path.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>

Deedone committed:
drivers: tee: optee: Implement SHM_TYPE_APPL allocation

Form and pass the allocation requests to the supplicant when
SHM_TYPE_APPL is requested by the OPTEE. Register allocated region in
the OPTEE afterwards.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>

Deedone committed:
tests: drivers: optee: Unit tests for supplicant alloc

Basic tests for supplicant shared memory allocation.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>

oleksiimoisieiev commented:
Compliance check was failed

oleksiimoisieiev commented on drivers/tee/optee/optee.c:
```diff
@@ -59,11 +59,34 @@ struct optee_notify {
 };
 
+struct optee_supp_req {
+	sys_dnode_t link;
```
s/syscals/syscalls/g in commit description

oleksiimoisieiev commented on drivers/tee/optee/optee.c:
```diff
@@ -205,6 +228,45 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
 
 static void *optee_construct_page_list(void *buf, uint32_t len, uint64_t *phys_buf);
 
+uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t num_params,
```
static maybe? Is it used from outside?

oleksiimoisieiev commented on drivers/tee/optee/optee.c:
```diff
@@ -205,6 +228,45 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
+			struct tee_param *param)
+
+{
+	struct optee_driver_data *data = (struct optee_driver_data*) dev->data;
```
should be space between ) and dev->data? I'm not sure, please consult with checkpatch

oleksiimoisieiev commented on drivers/tee/optee/optee.c:
```diff
@@ -449,6 +511,31 @@ static void free_shm_pages(void **pages)
+static void handle_rpc_supp_cmd(const struct device *dev, struct optee_msg_arg *arg)
+{
+	// TODO: FIX HARDCODE
+	struct tee_param params[5];
```
Hm. It's not clear to me why have you made this hardcode? IIUC we have num_params passed in arg argument.
At least if there is a good reason why have you hardcoded this param I'd like to see it in TODO.

oleksiimoisieiev commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +961,95 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+	}
+	req->ret = ret;
+
+	supp->current = NULL;
```
This probably should be under the mutex as well. Maybe it's better to make supp->current= NULL right after reading it's value so revc won't get BUSY?

oleksiimoisieiev commented on tests/drivers/tee/optee/src/main.c:
```diff
@@ -470,6 +470,149 @@ ZTEST(optee_test_suite, test_func_shm_alloc)
+
+	ret = tee_suppl_recv(dev, &func, &num_params, params);
+
+	if (ret != 0) {
```
why not using z_assert here to check return values?

Deedone replied:
It causes the main thread to hang when assert fails.

oleksiimoisieiev replied:
+1 let's leave as is

oleksiimoisieiev commented on drivers/tee/optee/optee.c:
```diff
@@ -282,19 +316,17 @@ static void handle_cmd_alloc(const struct device *dev, struct optee_msg_arg *arg
-	 */
-	if (arg->params[0].u.value.a != OPTEE_RPC_SHM_TYPE_KERNEL) {
+	if (arg->params[0].u.value.a == OPTEE_RPC_SHM_TYPE_KERNEL) {
+		/* todo handle situation when shm was allocated statically so buffer can be reused*/
```
s/todo/TODO/g

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -59,11 +59,34 @@ struct optee_notify {
+	size_t num_params;
+	struct tee_param *param;
+
+	struct k_sem c;
```
Strange naming, purpose is not clear.

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -156,7 +179,7 @@ static int msg_param_to_param(struct tee_param *param, unsigned int num_param,
 		case OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT:
 		case OPTEE_MSG_ATTR_TYPE_VALUE_INOUT:
-			tp->attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT + attr -
+			tp->attr = TEE_PARAM_ATTR_TYPE_VALUE_INPUT + attr -
```
It should be mentioned in commit message or be a separate commit

Deedone replied:
Moved to a separate commit

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -205,6 +228,45 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
+	uint32_t ret;
+
+	req = k_malloc(sizeof(*req));
+	if (!req)
```
Add braces {} for all one-line blocks (checpatch.pl should be used with `--ignore BRACES`)

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +961,95 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+	while (true) {
+		k_mutex_lock(&supp->mutex, K_FOREVER);
+		if (!sys_dlist_is_empty(&supp->reqs)) {
+			req = (struct optee_supp_req *)sys_dlist_get(&supp->reqs);
```
redundant empty line

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +961,95 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+
+	*func = req->func;
+	*num_params = req->num_params;
+	memcpy(param, req->param,
```
may be a single line

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -282,19 +316,17 @@ static void handle_cmd_alloc(const struct device *dev, struct optee_msg_arg *arg
-	 * Currently TYPE_KERNEL was implemented.
-	 */
-	if (arg->params[0].u.value.a != OPTEE_RPC_SHM_TYPE_KERNEL) {
+	if (arg->params[0].u.value.a == OPTEE_RPC_SHM_TYPE_KERNEL) {
```
Maybe switch/case?

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -323,27 +355,27 @@ static void handle_cmd_alloc(const struct device *dev, struct optee_msg_arg *arg
-	 * Currectly TYPE_KERNEL was implemented.
-	 */
-	if (arg->params[0].u.value.a != OPTEE_RPC_SHM_TYPE_KERNEL) {
+	if (arg->params[0].u.value.a == OPTEE_RPC_SHM_TYPE_KERNEL) {
```
same about switch/case

Deedone committed:
drivers: optee: Fix typo in attr calculation

Change INOUT to INPUT to align the calculation with the rest of the
logic.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>

oleksiimoisieiev commented on drivers/tee/optee/optee.c:
```diff
@@ -476,8 +571,8 @@ static uint32_t handle_func_rpc_call(const struct device *dev, struct tee_shm *s
 		return TEEC_ERROR_NOT_IMPLEMENTED;
 	default:
 		/* TODO: supplicant commands are still not implemented */
```
Probably no use for this todo because commands were implemented.

oleksiimoisieiev commented:
LGTM after removing redundand TODO
`Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -449,6 +512,38 @@ static void free_shm_pages(void **pages)
+	if (ret) {
+		arg->ret = TEEC_ERROR_GENERIC;
+		arg->ret_origin = TEEC_ORIGIN_COMMS;
+		goto out;
```
nit: no need for this goto

Deedone replied:
Let's leave it in case someone adds some new code.

xakep-amatop replied:
someone can add `goto` if he or she add a new chunk of code

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -268,6 +268,40 @@ static uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t
+
+	ret = optee_call_supp(dev, OPTEE_RPC_CMD_SHM_ALLOC, 1, &param);
+
+	if (ret)
```
braces

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -268,6 +268,40 @@ static uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t
+{
+	struct tee_param param;
+
+
```
empty line

firscity commented on drivers/tee/optee/optee.c:
```diff
@@ -268,6 +268,40 @@ static uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t
+	struct tee_param param;
+
+
+	param.attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT;
```
may be done in initialization:
`... param = {
    .attr = ...
};`

Deedone replied:
I think it looks better with separate declaration and initialization.

xakep-amatop replied:
note: here can be a leak of some data, also if we have more fields that you initialize here some of fields will contain garbage and it can influence working of OPTEE OS

initializers are like: `{}`, `{0}` and `{.attr = ...}` solve the last issue, but they still have security issues in case when we have paddings between fields in structure, they can contain some stack garbage and it can be useful for some hacker's attacks

so, the best way is to use here `memset` for `param`
but I don't see how somebody can use stack garbage in structure paddings in this case

firscity commented on tests/drivers/tee/optee/src/main.c:
```diff
@@ -617,6 +617,180 @@ ZTEST(optee_test_suite, test_func_rpc_supp_cmd)
+		goto fail1;
+	}
+
+	mem = k_malloc(params[0].b);
```
check return?

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+	struct optee_supp_req *req;
+	size_t n;
+
+	if (!supp->current)
```
I believe that we can drop this check

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
 }
 
-static int optee_suppl_send(const struct device *dev, unsigned int num_params,
+static int optee_suppl_send(const struct device *dev, unsigned int ret, unsigned int num_params,
```
do we really need `num_params`? should you check that we have enough number of items to hold `req->num_params`
as far as I see, it is unused

Deedone replied:
Added check

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+	}
+
+	k_mutex_lock(&supp->mutex, K_FOREVER);
+	supp->current = req;
```
can be a situation when we still have another valid/non-handled  `current`?
could you please add assert here at least

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req = NULL;
+
+	if (supp->current) {
```
looks like, it isn't safe
what if somebody sets `supp->current` right after this check from another thread?

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+	}
+
+	if (*num_params < req->num_params) {
+		return -EINVAL;
```
should we process all list items in order or if we have enough space in `param` array for another item from list we can process it right now ? 
in any case, could you please add a few words about it to the code ?

xakep-amatop replied:
should you put back to the list `req` or, at least, notify the waiting thread about completion??

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+		k_sem_take(&supp->reqs_c, K_FOREVER);
+	}
+
+	if (*num_params < req->num_params) {
```
isn't better to move this check and assignment to the above loop?
I mean is it good idea to break mutex lock? why can it be done under one lock in one place?

xakep-amatop reviewed (commented):
the review hasn't been finished yet
I've posted some questions just to speed up addressing of issues if there are any

xakep-amatop commented:
only for two next commits:  538c0381ba385fe51eb9b0aab92889a57ca66f02 and b6485303df678d5e5c1594be61e7b65c05371c5f
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>`

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +967,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+
+	if (num_params != req->num_params) {
+		LOG_ERR("Invalid number of parameters, expected %lu got %u", req->num_params,
+			num_params);
```
should you add here the next chunk of code:
```
req->ret = -EINVAL;
k_sem_give(&req->complete);
```
??

I think you should to wake up the waiting thread, but I'm not sure about the error code setting like above
or caller should make this call again with bigger params array

Deedone replied:
An invalid number of parameters here indicates that there is a response for some other request, so I think it's better to just ignore it.

xakep-amatop replied:
who will notify the waiting thread in that case?

Deedone replied:
It will be notified when a response with the correct number of parameters arrives.

xakep-amatop replied:
how is it not possible? where can you get ptr to req?
as far as I see, you erased `supp->current` a few lines above

so, req not in queue and not in `supp->current`

Deedone replied:
You are right, didn't catch that.

firscity reviewed (approved):
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +967,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+
+		if (req && supp->current == NULL) {
+			if (*num_params < req->num_params) {
+				LOG_ERR("Not enough space for params, need at least %lu",
```
it is up to you, because I don't know what will be better
but I think in this case caller should alloc more params and call this function again or you can drop entry from the list and notify the waiting thread

Deedone replied:
With this new logic, the request is left in the queue in case of error, awaiting correct retrieval by the supplicant.

xakep-amatop replied:
ok, got it, I hope it doesn't cause the stall

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -874,15 +967,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
+		k_mutex_lock(&supp->mutex, K_FOREVER);
+		req = (struct optee_supp_req *)sys_dlist_peek_head(&supp->reqs);
+
+		if (req && supp->current == NULL) {
```
so, can be a situation when you have `req != NULL` and `supp->current != NULL`
than the below check:
```
if (req)
```
finishes the loop 

is it what we want ?

sa-kib commented on drivers/tee/optee/optee.c:
```diff
@@ -205,6 +228,80 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
+	param.b = (uint64_t)shm;
+	param.c = 0;
+
+	optee_call_supp(dev, OPTEE_RPC_CMD_SHM_FREE, 1, &param);
```
```suggestion
	optee_call_supp(dev, OPTEE_RPC_CMD_SHM_FREE, 1, &param);
	tee_rm_shm(dev, shm);
```

we must free `shm` allocated in `cmd_alloc_suppl()`

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -205,6 +228,46 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
+	sys_dlist_append(&supp->reqs, &req->link);
+	k_mutex_unlock(&supp->mutex);
+
+	/* Tell an eventual waiter there's a new request */
```
"event listener"

"Eventual waiter" is something completely different : )

lorc commented on include/zephyr/drivers/tee.h:
```diff
@@ -349,7 +349,7 @@ typedef int (*tee_shm_unregister_t)(const struct device *dev, struct tee_shm *sh
  * See @a tee_suppl_recv() for argument definitions.
  */
-typedef int (*tee_suppl_recv_t)(const struct device *dev, uint32_t func, unsigned int num_params,
+typedef int (*tee_suppl_recv_t)(const struct device *dev, uint32_t *func, unsigned int *num_params,
```
Changes in this file should go into a separate commit with a note that they should be squashed into original commit during upstreaming process

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -876,15 +1011,101 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
 {
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req;
+	size_t n;
+
+	k_mutex_lock(&supp->mutex, K_FOREVER);
+	req = supp->current;
+	k_mutex_unlock(&supp->mutex);
+
+	if (!req) {
+		return -EINVAL;
+	}
+
+	if (num_params != req->num_params) {
+		LOG_ERR("Invalid number of parameters, expected %lu got %u", req->num_params,
+			num_params);
+		return -EINVAL;
+	}
```
```suggestion
	struct optee_supp_req *req = NULL;
	size_t n;

	k_mutex_lock(&supp->mutex, K_FOREVER);
	if (supp->current && num_params >= supp->current->num_params) {
		req = supp->current;
		supp->current = NULL;
	} else {
		LOG_ERR("Invalid number of parameters, expected %lu got %u", req->num_params,
			num_params);
	}
	k_mutex_unlock(&supp->mutex);

	if (!req) {
		return -EINVAL;
	}
```

xakep-amatop reviewed (approved):
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>` commit 21d89b4a45af7f14e5daf9959aa683a641d5422b
I have finished review

Deedone committed:
drivers: optee: Change signatures for supplicant send/recv

Change tee_suppl_recv signature to allow for returning of func and
num_params variables.

Change tee_supl_send signature to allow supplicant to provide the return
code.

This commit should be squashed with
f953fc5a466dfa54a0be2a234e6e7458d2515c8a
drivers: tee: Add Tee driver generic API
before upstreaming.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>

Deedone committed:
drivers: tee: optee: Implement supplicant communication

Implement suppl_send/suppl_receive syscalls and optee_call_supp function.

Requests passed to optee_call_supp are forwarded to the supplicant, the
calling thread gets blocked while the request is processing.

Signed-off-by: Mykyta Poturai <mykyta_poturai@epam.com>
Acked-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>

lorc reviewed (approved):
`Acked-by: Volodymyr Babchuk <volodymyr_babchyk@epam.com>`
